rfcs/0128-selective-auto-merge-on-bot-upgrades.md

```diff
@@ -0,0 +1,56 @@
+---
+feature: Selective auto-merge on bot upgrades
+start-date: 2022-07-07
+author: superherointj
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a `meta.autoMerge` attribute to packages to allow committers to delegate merge rights to package maintainers for bot auto updates.
+
+# Motivation
+[motivation]: #motivation
+
+* Reduce pending PRs for review and merge in nixpkgs.
+* Save reviewers and commiters time.
+* Speed up package upgrades.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new `meta.autoMerge` package attribute with type `bool` defaulting to `false`. To be documented in Nixpkgs manual.
+
+OfBorg reads `meta.autoMerge` value and sets `13.auto-merge` label to PR.
+
+GitHub Actions will trigger a merge when the following 4 conditions are met:
+1) `@r-ryantm` is PR's author.
+2) Package `meta.autoMerge` attribute is enabled.
+3) All CI checks passed.
+4) Package maintainer has approved PR.
```

Comment:
How would that work if a package has multiple maintainers? Would they all need to approve? Would one be enough? Would 50% need to approve? Would we need to prune old maintainers?

Reply:
Currently a single maintainer approving would trigger autoMerge.
The policy can be improved if there is need.
Like: `meta.maintainerApprovalsForAutoMerge = 2;`
Or any other criteria that is desirable.

Regarding someone that is listed as a maintainer and is not wanted to be a maintainer, I guess, that is the case of removing the maintainer. An inactive maintainer wouldn't be approving PRs anyway.
The policy would be up to the commiter, he would have to consider which/the list of maintainers he is granting the permission of auto-merging bot PRs for the specific package.

Would be possible to have a list of maintainers that have the autoMerge permission and check for that. Like:
> `meta.maintainersWithAutoMerge = [];`

As an initial proposal, I'd prefer to keep this simple for a prototype at least.

Today we have no step between reviewer and committer. Granting limited permission in specific packages to merge PRs can be a step stone before becoming a commiter.

Reply:
> How would that work if a package has multiple maintainers? Would they all need to approve? Would one be enough? Would 50% need to approve?

Currently a single maintainer approving would trigger autoMerge.

The policy can be improved to fullfill needs. Like:

  `meta.maintainerApprovalsForAutoMerge = 2;`

  `meta.maintainersGrantedAutoMerge = [ maintainer1 maintainer2 ];`

The commiter would have to consider to whom and to which package he is granting auto-merge permission.

> Would we need to prune old maintainers?

Inactive maintainers don't approve PRs. => Not an issue?

Today we have no step between reviewer and committer. Granting a limited permission (bot upgrades) in a specific package to merge PRs can be a step stone before becoming a commiter.

As an initial proposal, I'd prefer to keep this simple for a prototype at least.

Reply:
I would default to 50% of the maintainers listed.